### PR TITLE
AArch64: Implement EDO

### PR DESCRIPTION
**Pull request by knn-k**
This commit implements EDO (exception directed optimization) for AArch64.

### CODE REVIEW

**Comment by knn-k** [comment]
Jenkins test sanity amac jdk17

**Comment by knn-k** [comment]
Jenkins test extended.functional alinux64 jdk17

**Comment by knn-k** [comment]
Added the call to `setHasEdoSnippet()`.

**Comment by knn-k** [comment]
Jenkins test sanity alinux64,amac jdk11

**Comment by knn-k** [comment]
Fixed the merge conflicts.

**Comment by knn-k** [comment]
Jenkins test extended.functional alinux64 jdk17

**Comment by knn-k** [comment]
Jenkins test sanity amac jdk17

File: runtime/compiler/aarch64/codegen/ForceRecompilationSnippet.cpp [hunk 1]
```diff
@@ -20,31 +20,129 @@
...
+   startPCReg->setRegisterFieldRD((uint32_t *)cursor);
+   cursor += ARM64_INSTRUCTION_LENGTH;
+
+   intptr_t helperAddress = (intptr_t)induceRecompilationSymRef->getMethodAddress();
+   intptr_t distance = helperAddress - (intptr_t)cursor;
+   if (!constantIsSignedImm28(distance))
+      {
+      helperAddress = TR::CodeCacheManager::instance()->findHelperTrampoline(induceRecompilationSymRef->getReferenceNumber(), (void *)cursor);
+      distance = helperAddress - (intptr_t)cursor;
+      TR_ASSERT(constantIsSignedImm28(distance), "Trampoline too far away.");
```
[the remark this answers is not in the document]
Comment (knn-k) [reply runtime/compiler/aarch64/codegen/ForceRecompilationSnippet.cpp]: Fixed.

File: runtime/compiler/aarch64/codegen/ForceRecompilationSnippet.cpp [hunk 2]
```diff
@@ -20,31 +20,129 @@
...
+      distance = (intptr_t)(_restartLabel->getCodeLocation()) - (intptr_t)cursor;
+      if (constantIsSignedImm28(distance))
+         {
+         // b distance
+         *(uint32_t *)cursor = TR::InstOpCode::getOpCodeBinaryEncoding(TR::InstOpCode::b) | ((distance >> 2) & 0x3ffffff); // imm26
+         cursor += ARM64_INSTRUCTION_LENGTH;
+         }
+      else
+         {
+         TR_ASSERT(false, "Target too far away.  Not supported yet");
```
[the remark this answers is not in the document]
Comment (knn-k) [reply runtime/compiler/aarch64/codegen/ForceRecompilationSnippet.cpp]: Fixed.

File: runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp [hunk 3]
```diff
@@ -105,7 +105,45 @@ void VMgenerateCatchBlockBBStartPrologue(TR::Node *node, TR::Instruction *fenceI
...
+      generateTrg1MemInstruction(cg, TR::InstOpCode::ldrimmx, node, recompCounterReg, loadbiMR);
+      generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::subsimmx, node, recompCounterReg, recompCounterReg, 1);
+      generateMemSrc1Instruction(cg, TR::InstOpCode::strimmx, node, storebiMR, recompCounterReg);
+      cg->stopUsingRegister(biAddrReg);
+      cg->stopUsingRegister(recompCounterReg);
+
+      TR::LabelSymbol *snippetLabel = generateLabelSymbol(cg);
+      TR::LabelSymbol *doneLabel = generateLabelSymbol(cg);
+
+      // Registers killed by the call through jitCallCFuntion
```
[the remark this answers is not in the document]
Comment (knn-k) [reply runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp]: Fixed.

**Comment by 0xdaryl** [comment]
Jenkins test sanity alinux64 jdk17

**Pull request merged**